dali/kernels/reduce/reduce_all_kernel_gpu.h

```diff
@@ -0,0 +1,107 @@
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/reduce/reduce.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce {
+
+template <typename T, int ndim, typename Reductor, typename Preprocessor = identity>
+class DLL_PUBLIC ReduceAllGPU {
+ public:
+  // TODO(janton): remove this when implemented
+  static_assert(std::is_same<Preprocessor, identity>::value,
+                "Preprocessing is not yet implemented");
+
+  DLL_PUBLIC ~ReduceAllGPU() = default;
+
+  DLL_PUBLIC KernelRequirements Setup(KernelContext &context,
+                                      const InListGPU<T, ndim> &in) {
+    auto num_samples = in.size();
+    ScratchpadEstimator se;
+
+    int64_t max_sample_size = 0;
+    for (int i = 0; i < num_samples; i++) {
+      max_sample_size = std::max(max_sample_size, volume(in.tensor_shape(i)));
+    }
+    blocks_per_sample_ = std::max(32, static_cast<int>(std::sqrt(max_sample_size)));
+
+    se.add<const T*>(AllocType::Host, num_samples);
+    se.add<int64_t>(AllocType::Host, num_samples);
+    se.add<const T*>(AllocType::GPU, num_samples);
+    se.add<int64_t>(AllocType::GPU, num_samples);
+    tmp_buffer_size_ = num_samples * blocks_per_sample_;
+    se.add<T>(AllocType::GPU, tmp_buffer_size_);
+
+    KernelRequirements req;
+    req.scratch_sizes = se.sizes;
+    req.output_shapes = {TensorListShape<1>::make_uniform(num_samples, TensorShape<1>{1})};
+    return req;
+  }
+
+  DLL_PUBLIC void Run(KernelContext &context,
+                      const OutListGPU<T, 1> &out,
+                      const InListGPU<T, ndim> &in) {
+    auto num_samples = in.size();
+    auto* sample_data = context.scratchpad->Allocate<const T*>(AllocType::Host, num_samples);
+    auto* sample_size = context.scratchpad->Allocate<int64_t>(AllocType::Host, num_samples);
+    for (int i = 0; i < num_samples; i++) {
+      sample_data[i] = in.tensor_data(i);
+      sample_size[i] = volume(in.tensor_shape(i));
+    }
+
+    auto* sample_data_gpu = context.scratchpad->Allocate<const T*>(AllocType::GPU, num_samples);
+    auto* sample_size_gpu = context.scratchpad->Allocate<int64_t>(AllocType::GPU, num_samples);
+    // Single memcpy, since the data in the scratchpad is contiguous
+    cudaMemcpyAsync(sample_data_gpu, sample_data,
+                    num_samples * sizeof(const T*) + num_samples * sizeof(int64_t),
+                    cudaMemcpyHostToDevice, context.gpu.stream);
+
+    auto* buffer_gpu = context.scratchpad->Allocate<T>(AllocType::GPU, tmp_buffer_size_);
+
+    dim3 block(32, 32);
+    dim3 grid(blocks_per_sample_, num_samples);
+    ReduceAllBatchedKernel<<<grid, block>>>(buffer_gpu, sample_data_gpu, sample_size_gpu,
+                                            reduction);
+
+    DALI_ENFORCE(out.is_contiguous(), "Reduce all kernel expects the output to be contiguous");
```

Comment:
Move this test to the top - there's no point starting work when this is not satisfied.

Reply:
done